### PR TITLE
HoloHash from functions avoid panicking unless desired

### LINKED ISSUE #3313: Remove panicky `HoloHash::from_raw_39` and modify downstream consumers to also not panic
**Issue**
Calling e.g. `AgentPubKey::from_raw_39(bytes)` on something that isn't 39 bytes long causes an assertion failure. This seems surprising seeing as the function returns a `HoloHashResult`, which includes the variant `Err(HoloHashError::BadSize)`. This forces the developer to check the length proactively in addition to handling the result.

**Expected behaviour**
The function returns an `Err` instead of panicking.

### PR DESCRIPTION
**Pull request by mattyg**
### Summary
Partially addresses #3313 

- **BREAKING** Renamed `from_raw_39_panicky` to `from_raw_39`.
- Added function `try_from_raw_39` which returns an Err instead of panicking.
- Added function `try_from_raw_36_and_type` which returns an Err instead of panicking.
- Happy and error tests for all 4 functions.

Not sure if I should avoid the breaking change and keep `from_raw_39_panicky` as an alias to `from_raw_39`

### TODO:
- [x] CHANGELOGs updated with appropriate info
- [x] All code changes are reflected in docs, including module-level docs

### CODE REVIEW

File: crates/holo_hash/src/hash.rs [hunk 1]
```diff
@@ -116,25 +116,37 @@ where
...
-    pub fn from_raw_36_and_type(mut bytes: Vec<u8>, hash_type: T) -> Self {
-        assert_length!(HOLO_HASH_UNTYPED_LEN, &bytes);
+    /// along with a type, to construct a hash.
+    pub fn try_from_raw_36_and_type(mut bytes: Vec<u8>, hash_type: T) -> HoloHashResult<Self> {
+        if bytes.len() != HOLO_HASH_UNTYPED_LEN {
+            return Err(HoloHashError::BadSize);
+        }
         let mut hash = hash_type.get_prefix().to_vec();
         hash.append(&mut bytes);
-        assert_length!(HOLO_HASH_FULL_LEN, &hash);
```
Comment (mattyg) [inline crates/holo_hash/src/hash.rs:136]: I don't see why we need to check the length again.

File: crates/holochain_state/src/query/tests/links_test.rs
Comment (mattyg) [inline crates/holochain_state/src/query/tests/links_test.rs:1]: This is all unrelated clippy fixes

File: crates/holo_hash/src/hash.rs [hunk 2]
```diff
@@ -116,25 +116,37 @@ where
...
         Ok(Self { hash, hash_type })
     }
+
     /// Raw constructor: Create a HoloHash from 39 bytes, using the prefix
-    /// bytes to determine the hash_type. Panics if hash_type does not match.
-    pub fn from_raw_39_panicky(hash: Vec<u8>) -> Self {
-        Self::from_raw_39(hash).expect("the specified hash_type does not match the prefix bytes")
+    /// bytes to determine the hash_type.
+    /// Panics if hash_type does not match or hash is incorrect length.
+    pub fn from_raw_39(hash: Vec<u8>) -> Self {
```
Comment (ThetaSinner) [inline crates/holo_hash/src/hash.rs:130]: This pattern with `from_xxx` and `try_from_xxx` suggests to me that in some way there is an infallible version of this operation and one where errors can happen. I'd prefer to just break the API here and have `from_xxx` return an error. I don't see why we'd ever want to panic when doing this, especially as we get these hashes as external inputs

**Comment by ThetaSinner** [comment]
I've been meaning to get to this for ages, thank you for picking it up! 

I'd quite like to make a more extreme change, because I don't think there was any real logic behind unwrapping here except to discover bugs in the Holochain code as it was being written. But that's not really relevant any more so I can't see why we'd want the versions that panic

File: crates/holo_hash/src/hash.rs (hunk 2 above)
Comment (ThetaSinner) [reply crates/holo_hash/src/hash.rs:130]: Yep please, that sounds good to me!

**Review by jost-s — APPROVED** [review]
I agree that there's no need for a panicking from. No harm either, because it's common in Rust to panic in case of failure in `from`s. No preference whether we keep or remove them.

**Comment by mattyg** [comment]
I suggest we merge this as-is and then remove the from_* functions in a subsequent PR. 

There are several layers of downstream functions using these which would also need to be refactored to return errors instead of panic so the surface area is much larger.
